pkg/lint/rules/template.go

```diff
@@ -3,39 +3,135 @@ package rules
+	// Load chart and parse templates, based on tiller/release_server
+	chart, err := chartutil.Load(linter.ChartDir)
+
+	if err != nil {
+		fmt.Errorf("templates: can not load Chart directory")
+		return
+	}
+
+	overrides := map[string]interface{}{
+		"Release": map[string]interface{}{
+			"Name":    "testRelease",
+			"Service": "Tiller",
+		},
+		"Chart": chart.Metadata,
+	}
+
+	chartValues, _ := chartutil.CoalesceValues(chart, chart.Values, overrides)
+	renderedContent, err := engine.New().Render(chart, chartValues)
+
+	renderOk := linter.RunLinterRule(support.ErrorSev, validateCanRender(err))
+
+	if !renderOk {
+		return
+	}
+
+	for _, template := range chart.Templates {
+		fileName, preExecutedTemplate := template.Name, template.Data
+
+		yamlFile := linter.RunLinterRule(support.ErrorSev, validateYamlExtension(linter, fileName))
+
+		if !yamlFile {
+			return
+		}
+
+		linter.RunLinterRule(support.WarningSev, validateQuotes(fileName, string(preExecutedTemplate)))
+
+		var yamlStruct K8sYamlStruct
+		// Even though K8sYamlStruct only defines Metadata namespace, an error in any other
+		// key will be raised as well
+		err := yaml.Unmarshal([]byte(renderedContent[fileName]), &yamlStruct)
+
+		validYaml := linter.RunLinterRule(support.ErrorSev, validateYamlContent(fileName, err))
+
+		if !validYaml {
+			return
+		}
+
+		linter.RunLinterRule(support.ErrorSev, validateNoNamespace(fileName, yamlStruct))
+	}
 }
 
-func validateTemplatesExistence(linter *support.Linter, templatesPath string) (lintError support.LintError) {
-	if _, err := os.Stat(templatesPath); err != nil {
+func validateTemplatesDir(linter *support.Linter, templatesPath string) (lintError support.LintError) {
+	if fi, err := os.Stat(templatesPath); err != nil {
 		lintError = fmt.Errorf("Templates directory not found")
+	} else if err == nil && !fi.IsDir() {
+		lintError = fmt.Errorf("'templates' is not a directory")
 	}
 	return
 }
 
-func validateTemplatesDir(linter *support.Linter, templatesPath string) (lintError support.LintError) {
-	fi, err := os.Stat(templatesPath)
-	if err == nil && !fi.IsDir() {
-		lintError = fmt.Errorf("'templates' is not a directory")
+// Validates that go template tags include the quote pipelined function
+// i.e {{ .Foo.bar }} -> {{ .Foo.bar | quote }}
+func validateQuotes(templateName string, templateContent string) (lintError support.LintError) {
+	r, _ := regexp.Compile(`(?m)\s{{([\w]|\.*|\s)+}}$`)
+	functions := r.FindAllString(templateContent, -1)
+
+	for _, str := range functions {
+		if match, _ := regexp.MatchString("quote", str); !match {
+			result := strings.Replace(str, "}}", " | quote }}", -1)
+			lintError = fmt.Errorf("templates: \"%s\". add \"| quote\" to your substitution functions: %s -> %s", templateName, str, result)
+			return
+		}
+	}
+	return
+}
+
+func validateYamlExtension(linter *support.Linter, fileName string) (lintError support.LintError) {
+	if filepath.Ext(fileName) != ".yaml" {
```

Comment:
Are we ok with not supporting `.yml` or should I add it to the list of valid formats?